near-sdk-macros/src/core_impl/code_generator/attr_sig_info.rs

```diff
@@ -191,10 +192,40 @@ impl AttrSigInfo {
                         }
                     }
                     BindgenArgType::CallbackResultArg => {
-                        let deserialize = deserialize_data(serializer_ty);
+                        let ok_type = extract_ok_type(ty).unwrap_or_else(||
+                            panic!("Function parameters marked with #[callback_result] should have \
+                                type Result<T, PromiseError>")
+                        );
+
+                        let deserialization_branch = match ok_type {
+                            // The unit type in this context is a bit special because functions
+                            // without an explicit return type do not serialize their response.
+                            // But when someone tries to refer to their callback result with
+                            // `#[callback_result]` they specify the callback type as
+                            // `Result<(), PromiseError` which cannot be correctly deserialized from
+                            // an empty byte array.
+                            //
+                            // So instead of going through serde, we do a manual check
+                            // that the byte array is either empty or contains the string "null".
+                            syn::Type::Tuple(type_tuple) if type_tuple.elems.is_empty() => {
+                                quote! {
+                                    near_sdk::PromiseResult::Successful(data) if data.is_empty() || data == vec!(110, 117, 108, 108) =>
+                                        Ok(()),
+                                    near_sdk::PromiseResult::Successful(_) =>
+                                        panic!("Failed to deserialize callback using JSON: expected\
+                                            an empty byte array or \"null\"")
```

Comment:
This fails when the Ok type is a tuple.

Reply:
Not sure what you mean here?

Reply:
Just noting that if the type was `Result<(X, Y), PromiseError>` or something like this, it would have failed by the previous logic. I believe it's fixed with the changes below, but I will double-check tomorrow

Reply:
Nah, I am checking that the tuple is empty so we only enter this branch when the tuple is `()`

Reply:
ah, I misread and had a bug in a test case I made. Sorry!